### DISCUSSION TITLE
Help understanding meaning of webpack chunk loading global state and empty chunks

### DISCUSSION THREAD

**Discussion by scameron** [webpack/webpack#19020] [Q&A]
Hello,

I'm looking at the global variable webpack uses for chunk loading and I'm hoping somebody can help me understand what I'm seeing.  The variable I'm referring to is the one whose name can (optionally) be defined by [output.chunkLoadingGlobal](https://webpack.js.org/configuration/output/#outputchunkloadingglobal).

There is an element in this global array for every chunk that is loaded.  Inside each element are 2 more arrays.  The first array to identify the chunk.  The second array to list all of the modules that live inside that chunk.

Normally, the "chunk id" array contains only a single element, with the name of the chunk in it.  But I found quite a number of cases in my build where the "chunk id" array has multiple elements, with each one referring to a different chunk.  For example, this one has 9 elements:

<img width="1238" alt="image" src="https://github.com/user-attachments/assets/934f59cc-95b5-4723-a09b-96edabf0ac0f">

If I open the chunk from the first element, `fieldSelection` I see that it indeed has this same 9 element array in it , and it contains all of the listed modules:

<img width="1614" alt="image" src="https://github.com/user-attachments/assets/bc21fc82-4302-47cf-aea6-4987273bb2d9">

My first question is:  what does it mean when a chunk has a list of other chunks in its id?  I'm guessing it means that all of these chunks were found to share the same list of modules?

Now, if I look at the build output on disk, I also find the other 8 chunks also exist (although they are not loaded into the browser at runtime).  If I open them, I see they are all basically empty.  They all look like this:

<img width="1524" alt="image" src="https://github.com/user-attachments/assets/0760d2e5-f351-4b93-9190-f9f983351a98">

Here you can see the chunk id list contains 8 elements instead of 9 elements.  The missing element is the `fieldSelection` one from above that had the actual modules in it.   All of the 8 changes are like this, with the 8 element chunk ids (differing only by order) and a single empty function with this same name `"?13bc"`.

Does anybody know why these nearly empty stump-chunks are being generated?  What is the empty function for?

In my webpack config I have `removeEmptyChunks` set to true.  Are these not considered empty chunks?  That `"?13bc"` function actually exists in quite a lot of my chunks, empty or not.

If anybody can shed some light on what this all means, I would really appreciate it.  I'm trying to reduce the size of my build and runtime footprint by reducing redundancy and duplicated modules, so understanding how this works will help me make progress in figuring the rest of that out.

**Comment by alexander-akait** [comment]
Looks like they are CSS chunks, right?

**Reply by scameron** [reply]
No, all the ones I've found like this so far are JS chunks.

**Reply by alexander-akait** [reply]
Unfortunately it is difficult to say without real code, if you make a reproducible minimal repository I will take a look

**Reply by scameron** [reply]
Yeah, that would be ideal but our build is enormous so it's quite difficult to simplify into a shareable repro case.

One new piece of information I learned today.  I had assumed that the `"?13bc"` module was some internal framework artifact, because of the name, but it turns out it's not.  I found a single reference to it inside `"./node_modules/htmlparser2/lib/WritableStream.js"`:

<img width="593" alt="image" src="https://github.com/user-attachments/assets/aa7eaa3d-beda-4ba4-b58b-9845c0b16593">

If I look the corresponding source file for that module, I can see that it is a require of `readable-stream`:

<img width="531" alt="image" src="https://github.com/user-attachments/assets/a57ea102-838d-4cc1-8b89-4f5146f0c557">

I have no idea why the module name for that module is `"?13bc"` instead of the actual module name, but that seems to be what it's referring to.

Does that ring any bells for you?  I found this old issue but not sure if it's related at all: https://github.com/webpack/webpack/issues/2342

**Reply by alexander-akait** [reply]
> I have no idea why the module name for that module is "?13bc" instead of the actual module name, but that seems to be what it's referring to.

Can you show your configuration? Especially externals and aliases, and also your target?

**Reply by scameron** [reply]
I tried to trim it down as much as I could, hopefully without removing anything relevant (just removed the stuff that's specific to our own code):

```
{
    amd: {
        jQuery: true
    },
    context: 'C:\\Example\\webpack',
    experiments: {
        css: true
    },
    externals: [{
        child_process: 'child_process'
    }],
    infrastructureLogging: {
        stream: {
            _eventsCount: 1,
            columns: 176,
            rows: 52,
            _type: 'tty',
            fd: 2,
            _isStdio: true
        },
        level: 'info',
        colors: true
    },
    mode: 'production',
    node: {
        global: true
    },
    optimization: {
        nodeEnv: 'production',
        usedExports: 'global',
        concatenateModules: true,
        minimize: true,
        removeAvailableModules: true,
        removeEmptyChunks: true,
        chunkIds: 'total-size',
        moduleIds: 'size',
        splitChunks: {
            chunks: 'all',
            minSize: 50000,
            minSizeReduction: 450000,
            minChunks: 1,
            maxAsyncRequests: 30,
            maxInitialRequests: 30
        }
    },
    output: {
        chunkFilename: 'uiAssets/app.chunk.[name].[chunkhash].js',
        chunkLoadingGlobal: 'webpackChunk_sac_api_w1732896845524',
        chunkLoadTimeout: 2400000,
        cssHeadDataCompression: true,
        crossOriginLoading: 'anonymous',
        filename: 'uiAssets/[name].main.[contenthash].js',
        path: 'C:\\Example\\webpack\\build',
        publicPath: 'auto',
        sourceMapFilename: 'uiAssets/sourcemaps/[filebase].map[query]'
    },
    resolve: {
        fallback: {
            path: 'C:\\Example\\webpack\\node_modules\\path-browserify\\index.js',
            timers: 'C:\\Example\\webpack\\node_modules\\timers-browserify\\main.js',
            os: 'C:\\Example\\webpack\\node_modules\\os-browserify\\browser.js',
            stream: 'C:\\Example\\webpack\\node_modules\\stream-browserify\\index.js',
            buffer: 'C:\\Example\\webpack\\node_modules\\buffer\\index.js',
            zlib: 'C:\\Example\\webpack\\node_modules\\browserify-zlib\\lib\\index.js',
            util: 'C:\\Example\\webpack\\node_modules\\util\\util.js',
            url: 'C:\\Example\\webpack\\node_modules\\url\\url.js',
            assert: 'C:\\Example\\webpack\\node_modules\\assert\\build\\assert.js'
        },
        extensions: ['.js', '.ts', '.tsx'],
        alias: {
            mathjs: 'C:\\Example\\webpack\\node_modules\\mathjs\\lib\\browser\\math',
            mongoose: 'C:\\Example\\webpack\\node_modules\\mongoose\\index',
            'antlr4/index': 'C:\\Example\\webpack\\node_modules\\antlr4\\index',
        }
    },
    stats: {
        assetsSpace: Infinity,
        colors: true
    },
    target: 'browserslist: last 1 Chrome versions, last 1 Edge versions, iOS >= 10.7'
}
```

**Reply by alexander-akait** [reply]
I think it is due `fallback` and `os-browserify`, unfortunately it is difficult to say without a real example

**Reply by scameron** [reply]
If I comment the fallback out, I get build errors.  So I tried to set `os` to false, and then I tried another build with all of the fallbacks set to false.  But in both cases I still get that weird module name showing up in my chunks with an empty function.

**Reply by alexander-akait** [reply]
If you try to bundle only htmlparser2 do you have the same problem?

**Reply by scameron** [reply]
Sorry, I'm not sure I understand.  What did you want me to try exactly?

**Reply by alexander-akait** [reply]
Import only `htmlparser2`

**Reply by scameron** [reply]
I've been banging my head against the wall trying to get a simple entry point to produce output but, for reasons I don't understand, I just can't get it to work.  If I use a dead simple entry point module with no dependencies, I don't get any output.  If I `require` another module, I still don't get output.  If I add a self-invoking function containing a dynamic `import` to one of our modules with significant code and other dependencies, then I get some output.  It must be something in our config, but I really don't know what's causing this.

So in the end, I'm not able to do a simple import of `htmlparser` and produce build output that I can check.

If you have any ideas what might be causing this, or have a suggestion for a simple entry point implementation that would work, let me know and I can try it out.

**Reply by alexander-akait** [reply]
@scameron disable minimize - https://webpack.js.org/configuration/optimization/#optimizationminimize

**Reply by scameron** [reply]
When I was trying to pinpoint the problem, I commented out the entire `optimization` config, but it didn't help.  As an experiment, I just brought it back with only `minimize` in it, set to false, but I still see the same behavior.

**Reply by alexander-akait** [reply]
Because this is due `fallback`, it looks like some modules are just empty,

try:

```
resolve: {
		alias: {
			"/style.js": false
		}
	},
```

```
import * as styles from "/style.js";
```	

And you will get the same output

**Reply by scameron** [reply]
Yeah, that makes sense.  I'm not sure exactly how it is happening in my particular case, but certainly there a lot of complicated aliases and fallbacks and weird exports defined in the packages of the dependencies like `readable_stream` and others.  So it's likely, as you say, that some modules just end up as empty.

But the root question is why, if webpack encounters an empty module, does it still put that module into a chunk and create a require reference to it.  The fact that it has that generated module name implies that there must be some special processing happening.  So why not just inline an empty function instead of treating the empty module as a real module?

**Reply by alexander-akait** [reply]
> So why not just inline an empty function instead of treating the empty module as a real module?

We do this, but only for ES modules, in some cases we simply delete them (sideEffect free), but to say exactly why in this case it stayed I will need the code. In some cases if it is an asynchronous chunk we need to leave it to put it in the array of loaded modules and understand that we do not need to load it again.
